### PR TITLE
8311647: Memory leak in Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_ttyname_1r

### PR DESCRIPTION
**Pull request by kspeeyu**
Allocated data on error path is deleted before return

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311647](https://bugs.openjdk.org/browse/JDK-8311647): Memory leak in Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_ttyname_1r (**Bug** - P4)

### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14832/head:pull/14832` \
`$ git checkout pull/14832`

Update a local copy of the PR: \
`$ git checkout pull/14832` \
`$ git pull https://git.openjdk.org/jdk.git pull/14832/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14832`

View PR using the GUI difftool: \
`$ git pr show -t 14832`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14832.diff">https://git.openjdk.org/jdk/pull/14832.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14832#issuecomment-1631172304)

### CODE REVIEW

**Review by phohensee — APPROVED** [review]

File: src/jdk.internal.le/linux/native/lible/CLibrary.cpp [hunk 1]
```diff
@@ -188,6 +188,7 @@ JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibr
 
     if (error != 0) {
         throw_errno(env);
+        delete[] data;
```
Comment (shipilev) [inline src/jdk.internal.le/linux/native/lible/CLibrary.cpp:191]: While JNI would not transfer control immediately on `Throw`, and would only throw it on return, it is a good style to do cleanups first, and then do throw, followed by immediate return.

In other words, flip the order of `delete[]` and `throw_errno`.

**Review by shipilev — CHANGES_REQUESTED** [review]

File: src/jdk.internal.le/linux/native/lible/CLibrary.cpp (hunk 1 above)
Comment (kspeeyu) [reply src/jdk.internal.le/linux/native/lible/CLibrary.cpp:191]: oh! got it

**Review by lahodaj — APPROVED** [review]
Looks good to me, thanks!

**Comment by shipilev** [comment]
Excellent, thanks!

/sponsor

**Review by phohensee — APPROVED** [review]

**Pull request closed without merging**